src/compaction_filter.h

```diff
@@ -0,0 +1,111 @@
+#pragma once
+
+#include "db_impl.h"
+#include "rocksdb/compaction_filter.h"
+
+#include <utility>
+
+namespace rocksdb {
+namespace titandb {
+
+class TitanCompactionFilter final : public CompactionFilter {
+public:
+  explicit TitanCompactionFilter(const TitanDBImpl *db,
+                                 const CompactionFilter *origin, uint32_t cf_id)
+      : titan_db_impl_(db), origin_filter_(origin), column_family_id_(cf_id) {}
+
+  const char *Name() const override { return "TitanCompactionFilter"; }
+
+  Decision FilterV2(int level, const Slice &key, ValueType value_type,
+                    const Slice &value, std::string *new_value,
+                    std::string *skip_until) const override {
+    if (value_type != kBlobIndex) {
+      return origin_filter_->FilterV2(level, key, value_type, value, new_value,
+                                      skip_until);
+    }
+
+    BlobIndex blob_index;
+    Slice origin_value(value.data());
+    Status s = blob_index.DecodeFrom(&origin_value);
+    if (!s.ok()) {
+      // Unable to decode blob index. Keeping the value.
```

Comment:
maybe should set background error here? What do you think? @yiwu-arbug

Reply:
What does **background error** mean? Write an error log? @Connor1996

Reply:
rocksdb/titan will enter read only mode once it hits write error or compaction error. There's this flag in titan: https://github.com/tikv/titan/blob/master/src/db_impl.h#L282.

But since the compaction filter doesn't provide API to return its error, I think it is also fine to return `kKeep`.